File: frontend/src/lib/textLimits.ts
/**
 * Längengrenzen für lange Freitexte — Karten-Beschreibungen und Kommentare (Issue #572).
 *
 * Gegenstück zu `org.mwolff.manban.common.TextLimits`. Der Wert steht im Frontend genau hier;
 * Eingabefelder und der Spezifikationsimport importieren ihn, statt ihn zu wiederholen.
 *
 * Zählweise: eine UTF-16-Codeeinheit, also `text.length` — dieselbe Einheit, die Bean Validation
 * über `@Size` prüft. Ein Emoji außerhalb der BMP zählt als zwei.
 */
export const MAX_TEXT_LENGTH = 50_000

/**
 * Fehlertext für ein zu langes Feld, z. B. „60.000 / 50.000 Zeichen".
 *
 * Bewusst ohne die Differenz: Sie ist aus beiden Zahlen ablesbar, und ein drittes Zahlenfeld im
 * Fehlertext wäre eine weitere Stelle, die auseinanderlaufen kann.
 */
export function tooLongMessage(length: number): string {
  const format = (value: number) => value.toLocaleString('de-DE')
  return `${format(length)} / ${format(MAX_TEXT_LENGTH)} Zeichen`
}

/** Ob der Text die Grenze reißt — dieselbe Prüfung für Feldzustand und Aktions-Sperre. */
export function isTooLong(text: string): boolean {
  return text.length > MAX_TEXT_LENGTH
}

File: src/main/java/org/mwolff/manban/common/TextLimits.java
package org.mwolff.manban.common;

/**
 * Längengrenzen für lange Freitexte — Karten-Beschreibungen und Kommentare (Issue #572).
 *
 * <p>Modulübergreifend, weil {@code card} und {@code comment} getrennte Module sind und dieselbe
 * Grenze tragen müssen: Zwei Konstanten mit demselben Zahlenwert laufen bei der nächsten Änderung
 * auseinander, und genau das war der Zustand davor — Kommentare waren auf 10.000 begrenzt,
 * Beschreibungen serverseitig gar nicht, und die Oberfläche kappte still bei 10.000.
 *
 * <p><strong>Zählweise:</strong> Ein „Zeichen" ist eine UTF-16-Codeeinheit — die Einheit, die
 * {@code String.length()} in Java und {@code String.prototype.length} in JavaScript ohne Zusatzcode
 * teilen, und die Bean Validation über {@code @Size} ohnehin prüft. Ein Emoji außerhalb der BMP
 * zählt damit als zwei. Das ist bewusst nicht die PostgreSQL-Zählweise ({@code length()} zählt
 * Codepoints); wer den Bestand gegen diese Grenze misst, nimmt {@code octet_length}, das als
 * UTF-8-Byte-Zahl immer mindestens so groß ist wie die UTF-16-Länge.
 */
public final class TextLimits {

  /**
   * Obergrenze für Beschreibungen und Kommentare.
   *
   * <p>Die frühere Grenze von 10.000 war an der Größenordnung „ein Block Fließtext" bemessen. Das
   * trug, solange Menschen die Texte tippten; seit der Tracker seine eigenen Issue-Bodies,
   * Reviewberichte und Abschlussberichte aufnimmt, sind fünfstellige Längen der Normalfall.
   */
  public static final int MAX_TEXT = 50_000;

  private TextLimits() {}
}
